ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs

```diff
@@ -725,3 +753,81 @@ data MiniProtocolFatalException = MiniProtocolFatalException
   deriving (Show)
 
 instance Exception MiniProtocolFatalException
+
+-----
+
+newtype PipeId = PipeId Word64
+  deriving (Enum, Eq, Ord, Show)
+
+-- | One half of a pair of connected 'Channel's: 'send' for one and 'recv' for
+-- the other
+--
+newtype Pipe m a = Pipe (Channel m a)
+
+-- | How many times has a pipe been emptied?
+--
+newtype EmptiedCount = EmptiedCount Word64
+  deriving (Enum, Eq, Ord, Show)
+
+data LivePipes a = LivePipes
+  { nextPipeId :: !PipeId
+  , livePipes  :: !(Map PipeId a)
+  }
+
+-- | A variable mapping each live pipe to its own variable, which contains
+-- @Just ec@ when the pipe is empty and @Nothing@ otherwise
+--
+type LivePipesVar m = StrictTVar m (LivePipes (StrictTVar m (Maybe EmptiedCount)))
+
+-- | Block until all live pipes have been empty for at least the given duration
+--
+blockUntilQuiescent :: IOLike m => LivePipesVar m -> DiffTime -> m ()
+blockUntilQuiescent livePipesVar dur = get >>= go
+  where
+    get = atomically $ do
+        LivePipes{livePipes} <- readTVar livePipesVar
+        forM livePipes $ \ecVar -> readTVar ecVar >>= \case
+            Nothing -> retry
+            Just ec -> pure ec
+
+    go sig1 = do
+      threadDelay dur
+      sig2 <- get
+      if sig1 == sig2 then pure () else go sig2
+
+-- | Create a pipe backed by a 'LazyTMVar', add it to the live pipes, and
```

Comment:
I only used `LazyTMVar` because that's what was being used. Since these mvars are simulating a network, it seems like making it strict would be in a sense more accurate. Though "time" is somewhat weird here due to `io-sim`, so I don't think it ultimately matters -- maybe only for tracing `error` calls perhaps?

I'll change to `StrictTMVar` and see if someone else objects.

Reply:
Ha, no, I think you're already using a `StrictTMVar`, since you're using `Control.Monad.Class.MonadSTM.Strict`. A partial type signature on `buffer` confirms this. So my comment only applies to the docstring :slightly_smiling_face:

Reply:
It now uses `TQueue`s. Though the function's comment is now pretty big and does not mention the exact data type.